plugins/inputs/intel_powerstat/intel_powerstat.go

```diff
@@ -1,915 +1,1218 @@
+// Stop deactivates perf events if one or more of the requested metrics rely on perf.
+func (p *PowerStat) Stop() {
+	if !p.needsPerf {
+		return
 	}
-}
 
-func (p *PowerStat) initRaplService() {
-	if p.packageCurrentPowerConsumption || p.packageCurrentDramPowerConsumption || p.packageThermalDesignPower || p.packageTurboLimit ||
-		p.packageUncoreFrequency || p.packageCPUBaseFrequency {
-		p.rapl = newRaplServiceWithFs(p.Log, p.fs)
+	if err := p.fetcher.DeactivatePerfEvents(); err != nil {
+		p.Log.Errorf("Failed to deactivate perf events: %v", err)
 	}
 }
 
-// fill CPUBusClockValue if required
-func (p *PowerStat) fillCPUBusClock() {
-	if p.packageCPUBaseFrequency {
-		// cpuBusClock is the same for every core/socket.
-		busClockInfo := p.getBusClock("0")
-		if busClockInfo == 0 {
-			p.Log.Warn("Disabling package metric: cpu_base_frequency_mhz. Can't detect bus clock value")
-			p.packageCPUBaseFrequency = false
-			return
-		}
+// Gather collects the plugin's metrics.
+func (p *PowerStat) Gather(acc telegraf.Accumulator) error {
+	// gather CPU metrics relying on coreFreq and msr which share CPU IDs.
+	if p.needsCoreFreq || p.needsMsrCPU {
+		p.addCPUMetrics(acc)
+	}
+
+	// gather CPU metrics relying on perf.
+	if p.needsPerf {
+		p.addCPUPerfMetrics(acc)
+	}
 
-		p.cpuBusClockValue = busClockInfo
+	// gather package metrics.
+	if len(p.PackageMetrics) != 0 {
+		p.addPackageMetrics(acc)
 	}
+
+	return nil
 }
 
-// Gather takes in an accumulator and adds the metrics that the Input gathers
-func (p *PowerStat) Gather(acc telegraf.Accumulator) error {
-	if p.areGlobalMetricsEnabled() {
-		p.addGlobalMetrics(acc)
+// parseConfig is a helper method that parses configuration fields from the receiver such as included/excluded CPU IDs.
+func (p *PowerStat) parseConfig() error {
+	var err error
+
+	if p.MsrReadTimeout < 0 {
+		return errors.New("msr_read_timeout should be positive number or equal to 0 (to disable timeouts)")
 	}
 
-	if p.areCoreMetricsEnabled() {
-		if p.msr.isMsrLoaded() {
-			p.logOnce["msr"] = nil
-			p.addPerCoreMetrics(acc)
-		} else {
-			err := errors.New("Error while trying to read MSR (probably msr module was not loaded)")
-			if val := p.logOnce["msr"]; val == nil || val.Error() != err.Error() {
-				p.Log.Errorf("%v", err)
-				// Remember that specific error occurs to omit logging next time
-				p.logOnce["msr"] = err
-			}
-		}
+	p.PackageMetrics, err = parsePackageMetrics(p.PackageMetrics)
+	if err != nil {
+		return fmt.Errorf("failed to parse package metrics: %w", err)
 	}
 
-	// Gathering the first iteration of metrics was skipped for most of them because they are based on delta calculations
-	p.skipFirstIteration = false
+	p.CPUMetrics, err = parseCPUMetrics(p.CPUMetrics)
+	if err != nil {
+		return fmt.Errorf("failed to parse core metrics: %w", err)
+	}
 
-	return nil
-}
+	if len(p.CPUMetrics) == 0 && len(p.PackageMetrics) == 0 {
+		return fmt.Errorf("no metrics were found in the configuration file")
```

Comment:
```suggestion
		return errors.New("no metrics were found in the configuration file")
```

Reply:
Done as part of commit [e2de6f6b](https://github.com/influxdata/telegraf/pull/14363/commits/e2de6f6b2577fe4e9ec7b420132b2750fe7e2654).